### PR TITLE
Do not morph gene on transfer but instead morph on mint

### PR DESCRIPTION
**Pull request by entrpyc**

### CODE REVIEW

File: contracts/Polymorph.sol [hunk 1]
```diff
@@ -58,23 +57,21 @@ contract Polymorph is IPolymorph, ERC721PresetMinterPauserAutoId, BMath, Reentra
...
-        if(whitelistAddresses[from] == true || whitelistAddresses[to] == true) {
-            ERC721PresetMinterPauserAutoId._beforeTokenTransfer(from, to, tokenId);
-            emit MarketplaceTransfer(tokenId, from, to);
-        } else {
-            uint256 oldGene = _genes[tokenId];
-            _genes[tokenId] = geneGenerator.random();
-            ERC721PresetMinterPauserAutoId._beforeTokenTransfer(from, to, tokenId);
-            emit TokenMorphed(tokenId, oldGene, _genes[tokenId]);
-        }
+        ERC721PresetMinterPauserAutoId._beforeTokenTransfer(from, to, tokenId);
```
Comment (Perseverance) [inline contracts/Polymorph.sol:60]: You can completely remove this function

File: contracts/Polymorph.sol [hunk 2]
```diff
@@ -58,23 +57,21 @@ contract Polymorph is IPolymorph, ERC721PresetMinterPauserAutoId, BMath, Reentra
...
     }
 
     function mint() public override payable nonReentrant {
         _tokenIdTracker.increment();
+
+        uint256 tokenId = _tokenIdTracker.current();
```
Comment (Perseverance) [inline contracts/Polymorph.sol:66]: Use the same variable in the _mint function instead of calling `current()` again

File: deployment/deploy.js [hunk 3]
```diff
@@ -3,12 +3,13 @@ const ethers = require('ethers');
...
-const deploy = async (network, secret, etherscanApiKey) => {
-
-	const deployer = new etherlime.InfuraPrivateKeyDeployer(secret, network, process.env.API_KEY);
-	deployer.defaultOverrides = { gasLimit: 4700000, gasPrice: 3000000000, etherscanApiKey };
-	const result = await deployer.deployAndVerify("etherscan", PolymorphWithGeneChanger, {}, "Polymorph", "MORPH", "http://www.kekdao.com/", "0xe42682eEa1DFC432C2fF5a779CD1D9a1e1c7f405", ethers.utils.parseEther("0.01"));
+const deploy = async (network, secret) => {
+	const daoAddress = '0xe42682eEa1DFC432C2fF5a779CD1D9a1e1c7f405';
+	const premintTokensCount = 5;
 
+	const deployer = new etherlime.InfuraPrivateKeyDeployer(secret, network, 'a40d4e5ac8764045953d6aa546c1e34d');
```
Comment (Perseverance) [inline deployment/deploy.js:10]: Never commit these.

**Review by Perseverance — CHANGES_REQUESTED** [review]

**Review by Perseverance — APPROVED** [review]

File: contracts/Polymorph.sol [hunk 4]
```diff
@@ -57,24 +56,25 @@ contract Polymorph is IPolymorph, ERC721PresetMinterPauserAutoId, BMath, Reentra
...
-        daoAddress.transfer(price);
-        _msgSender().transfer(msg.value.sub(price)); // Return excess
-        _mint(_msgSender(), _tokenIdTracker.current());
+
+        uint256 tokenId = _tokenIdTracker.current();
+        _genes[tokenId] = geneGenerator.random();
+
+        uint256 price = calcPolymorphPrice(tokenId);
+        
+        if(price > 0) {
```
Comment (strenev) [inline contracts/Polymorph.sol:67]: I think price will always be more than 0, so the check should be whether the excess amount is not 0.

File: contracts/PolymorphWithGeneChanger.sol [hunk 5]
```diff
@@ -28,8 +28,15 @@ contract PolymorphWithGeneChanger is IPolymorphWithGeneChanger, Polymorph {
     function morphGene(uint256 tokenId, uint256 genePosition) public payable virtual override nonReentrant {
         _beforeGenomeChange(tokenId);
         uint256 price = priceForGenomeChange(tokenId);
-        daoAddress.transfer(price);
-        _msgSender().transfer(msg.value.sub(price)); // Return excess
+        
+        if(price > 0) {
```
Comment (strenev) [inline contracts/PolymorphWithGeneChanger.sol:32]: Same as above

File: contracts/PolymorphWithGeneChanger.sol [hunk 6]
```diff
@@ -52,8 +59,15 @@ contract PolymorphWithGeneChanger is IPolymorphWithGeneChanger, Polymorph {
     function randomizeGenome(uint256 tokenId) public payable override virtual nonReentrant {
         _beforeGenomeChange(tokenId);
         uint256 price = priceForGenomeChange(tokenId);
-        daoAddress.transfer(price);
-        _msgSender().transfer(msg.value.sub(price)); // Return excess
+
+        if(price > 0) {
```
Comment (strenev) [inline contracts/PolymorphWithGeneChanger.sol:63]: Same as above

File: contracts/PolymorphWithGeneChanger.sol [hunk 7]
```diff
@@ -62,13 +76,15 @@ contract PolymorphWithGeneChanger is IPolymorphWithGeneChanger, Polymorph {
 
     function priceForGenomeChange(uint256 tokenId) public override virtual view returns(uint256 price) {
         uint256 pastChanges = _genomeChanges[tokenId];
+        require(pastChanges < 255, 'pastChanges is too big and will cause overflow.');
+
         price = baseGenomeChangePrice;
         
         for(uint256 i = 0; i < pastChanges; i++) {
```
Comment (strenev) [inline contracts/PolymorphWithGeneChanger.sol:83]: I think here the proposed implementation is without the for loop. Like this:
    
    function priceForGenomeChange(uint256 tokenId) public override virtual view returns(uint256 price) {
        uint256 pastChanges = _genomeChanges[tokenId];
        return baseGenomeChangePrice.mul(1 << pastChanges);
    }
    Can you test it and @Perseverance can confirm as well :)

**Pull request merged**